cache_flat_mutation_reader.hh

```diff
@@ -617,6 +634,72 @@ void cache_flat_mutation_reader::maybe_add_to_cache(const clustering_row& cr) {
     });
 }
 
+inline
+bool cache_flat_mutation_reader::maybe_add_to_cache(const range_tombstone_change& rtc) {
+    rows_entry::tri_compare q_cmp(*_schema);
+
+    clogger.trace("csm {}: maybe_add_to_cache({})", fmt::ptr(this), rtc);
+
+    // Don't emit the closing range tombstone change, we may continue from cache with the same tombstone.
+    // The following relies on !_underlying_upper_bound->is_clustering_row()
+    if (q_cmp(rtc.position(), *_underlying_upper_bound) == 0) {
+        _lower_bound = rtc.position();
+        return false;
+    }
+
+    auto prev = std::exchange(_current_tombstone, rtc.tombstone());
+    if (_current_tombstone == prev) {
+        return false;
+    }
+
+    if (!can_populate()) {
+        // _current_tombstone is now invalid and remains so for this reader. No need to change it.
+        _last_row = nullptr;
+        _population_range_starts_before_all_rows = false;
+        _read_context.cache().on_mispopulate();
+        return true;
+    }
+
+    _lsa_manager.run_in_update_section_with_allocator([&] {
+        mutation_partition_v2& mp = _snp->version()->partition();
+        rows_entry::tri_compare cmp(table_schema());
+
+        auto new_entry = alloc_strategy_unique_ptr<rows_entry>(
+                current_allocator().construct<rows_entry>(table_schema(), to_table_domain(rtc.position()), is_dummy::yes, is_continuous::no));
+        auto it = _next_row.iterators_valid() ? _next_row.get_iterator_in_latest_version()
+                                              : mp.clustered_rows().lower_bound(to_table_domain(rtc.position()), cmp);
+        auto insert_result = mp.mutable_clustered_rows().insert_before_hint(it, std::move(new_entry), cmp);
+        it = insert_result.first;
+        if (insert_result.second) {
+            _snp->tracker()->insert(*it);
+        }
+
+        rows_entry& e = *it;
+        if (ensure_population_lower_bound()) {
+            // underlying may emit range_tombstone_change fragments with the same position.
+            // In such case, the range to which the tombstone from the first fragment applies is empty and should be ignored.
+            if (q_cmp(_last_row.position(), it->position()) < 0) {
+                if (_read_context.is_reversed()) [[unlikely]] {
+                    clogger.trace("csm {}: set_continuous({}), rt={}", fmt::ptr(this), _last_row.position(), prev);
+                    _last_row->set_continuous(true);
+                    _last_row->set_range_tombstone(prev);
+                } else {
+                    clogger.trace("csm {}: set_continuous({}), rt={}", fmt::ptr(this), e.position(), prev);
+                    e.set_continuous(true);
+                    e.set_range_tombstone(prev);
+                }
+            }
```

Comment:
It may still be useful, we set _last_row to point to it. If we remove it, we won't be able to populate the next interval.

Reply:
if ensure_population_lower_bound() returns false, we cannot make it continuous, because (in the forward read case) the left-hand side of the interval is incomplete (it doesn't contain the last row).

Reply:
That's right, but I'm talking about the case when `ensure_population_lower_bound()  == true`, but  `q_cmp(_last_row.position(), it->position()) == 0`.

Reply:
In that case, we didn't really insert `e`, because there is already an entry with the same position. _last_row will still point to the old entry, and its continuity was set earlier when it was inserted. Here, we only change `_current_tombstone`.

Reply:
I see.